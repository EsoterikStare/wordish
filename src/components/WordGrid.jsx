import React from 'react';
import { Grid } from '@mui/material';

import Word from './Word';

import { useAppState } from '../AppStateContext';

const WordGrid = () => {
  const { state } = useAppState();

  // Empty words set based on current settings
  const getWordsTemplate = () => Array(state.maxGuesses).fill(
    Array(state.wordLength).fill({ guessChar: '' }),
  );

  // Empty words templates merged with previous guesses and current guess
  const getWords = () => getWordsTemplate().reduce((wordAcc, emptyWord, wordIndex) => {
    // Merge empty words with previous guesses and current guesses
    if (wordIndex < state.previousGuesses.length) {
      // Insert previous guesses at word indexes where those exist
      wordAcc.push(state.previousGuesses[wordIndex]);
    } else if (wordIndex === state.previousGuesses.length) {
      // Insert the characters of the current guess at the index after the last previous guess.
      const mergedGuess = emptyWord.reduce((mergeAcc, emptyChar, letterIndex) => {
        mergeAcc.push(state.currentGuess[letterIndex] || emptyChar);
        return mergeAcc;
      }, []);
      wordAcc.push(mergedGuess);
    } else {
      // Fill up the remaining words with empty placeholders
      wordAcc.push(emptyWord);
    }
    return wordAcc;
  }, []);

  return (
    <Grid data-testid="words-grid" alignItems="center" container item justifyContent="center" spacing={1}>
      {getWords().map((word, index) => (
        <Word key={`${word}-${index.toString()}`} word={word} />
      ))}
    </Grid>
  );
};

export default WordGrid;
