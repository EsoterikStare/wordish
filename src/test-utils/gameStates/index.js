import loseGameStateJSON from './loseGameState.json';
import midGuessGameStateJSON from './midGuessGameState.json';
import newGameStateJSON from './newGameState.json';
import winGameStateJSON from './winGameState.json';

export const loseGameState = JSON.stringify(loseGameStateJSON);
export const midGuessGameState = JSON.stringify(midGuessGameStateJSON);
export const newGameState = JSON.stringify(newGameStateJSON);
export const winGameState = JSON.stringify(winGameStateJSON);
