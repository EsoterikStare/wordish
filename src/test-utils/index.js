export * from './gameStates';
export { default as setup } from './setup';
