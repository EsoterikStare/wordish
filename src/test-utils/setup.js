import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { node } from 'prop-types';
import React from 'react';

import { AppStateProvider } from '../AppStateContext';
import AppThemeProvider from '../AppThemeProvider';

const Wrapper = ({ children }) => (
  <AppStateProvider>
    <AppThemeProvider>{children}</AppThemeProvider>
  </AppStateProvider>
);
Wrapper.propTypes = {
  children: node.isRequired,
};
Wrapper.defaultProps = {};

const setup = jsx => {
  const { rerender, ...rest } = render(<Wrapper>{jsx}</Wrapper>);
  return {
    user: userEvent.setup(),
    rerender: newJsx => rerender(<Wrapper>{newJsx}</Wrapper>),
    ...rest,
  };
};

export default setup;
