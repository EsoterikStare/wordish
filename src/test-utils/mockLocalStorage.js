import { newGameState } from './gameStates';

const mockLocalStorage = (gameState = newGameState) => {
  const originalLocalStorage = { ...window.localStorage };
  const mockValues = {
    clear: jest.fn(),
    // Provide a known win state from localStorage mock that will be used by AppStateProvider.
    getItem: jest.fn(() => gameState),
    removeItem: jest.fn(),
    setItem: jest.fn(),
  };

  // Mock window.localStorage with assertable functions.
  Object.defineProperty(window, 'localStorage', {
    value: mockValues,
    writable: true,
  });

  const resetLocalStorage = () => {
    Object.defineProperty(window, 'localStorage', {
      value: originalLocalStorage,
    });
  };

  return {
    resetLocalStorage,
    ...mockValues,
  };
};

export default mockLocalStorage;
