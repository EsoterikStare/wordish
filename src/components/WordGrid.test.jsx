import { screen } from '@testing-library/react';
import React from 'react';

import WordGrid from './WordGrid';

import { loseGameState, mockLocalStorage, setup } from '../test-utils';

describe('WordGrid', () => {
  it('should be defined', () => {
    expect(WordGrid).toBeDefined();
  });
  it('should render', () => {
    setup(<WordGrid />);
    expect(screen.getByTestId('words-grid')).toBeInTheDocument();
  });
  it('should correctly arrange Word components in a vertical stack', () => {
    mockLocalStorage(loseGameState);
    setup(<WordGrid />);
    expect(screen.getByTestId('words-grid')).toMatchSnapshot();
  });
});
