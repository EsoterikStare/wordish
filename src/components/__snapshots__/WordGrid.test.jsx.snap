// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordGrid should correctly arrange Word components in a vertical stack 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-8xdlsc-MuiGrid-root"
  data-testid="words-grid"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-1urxow7-MuiGrid-root"
    data-testid="word-grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        T
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        R
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        I
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-1eb3rk1"
        type="present"
      >
        A
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        L
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        S
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-1urxow7-MuiGrid-root"
    data-testid="word-grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        E
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-1eb3rk1"
        type="present"
      >
        N
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        O
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        U
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        G
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        H
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-1urxow7-MuiGrid-root"
    data-testid="word-grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        A
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        M
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        O
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        U
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        N
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        T
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-1urxow7-MuiGrid-root"
    data-testid="word-grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        A
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        C
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        O
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        U
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-7h3hx9"
        type="located"
      >
        N
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        T
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-1urxow7-MuiGrid-root"
    data-testid="word-grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        F
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        I
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        L
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        L
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        E
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        R
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-1urxow7-MuiGrid-root"
    data-testid="word-grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        F
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        I
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        L
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        L
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        E
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        R
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-1urxow7-MuiGrid-root"
    data-testid="word-grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        F
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        I
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        L
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        L
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        E
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      data-testid="letter-grid"
    >
      <div
        class="css-n6nm69"
        type="absent"
      >
        R
      </div>
    </div>
  </div>
</div>
`;
